### ISSUE TITLE
Different assemblies not in isolated AppDomain

### ISSUE THREAD

**Issue by cherrydev** [xunit/dnx.xunit#26]
Other/previous test runners ran each assembly in an isolated AppDomain.
When running tests for multiple assemblies/projects, each runs inside of the same AppDomain, at least as far as I can tell from `AppDomain.CurrentDomain.Id`.  Without this isolation, certain test scenarios become impossible to consistently replicate, such as tests that require different Entity Framework (6) DbConfigurations or anything that uses libraries that hold state in static fields.

**Comment by cherrydev** [comment]
After investigation, it appears that this is because xunit.execution.AspNet is being compiled with NO_APPDOMAIN.  Should I open a new issue on the core xunit project instead?

**Comment by bradwilson** [comment]
ASP.NET 5 does not support app domains. There is no way to support this.

**Comment by cherrydev** [comment]
Not strictly true.  The CoreCLR doesn't support app domains.  The regular CLR (which is part of ASP.NET 5) does.

**Comment by bradwilson** [comment]
Although the desktop .NET framework does support this, AFAIK the APIs are not exposed when targeting `aspnet50` (soon to be `dnx451`).

**Comment by bradwilson** [comment]
Another way to think about this is: if your only target will be `aspnet50`/`dnx451` (and not `aspnetcore50`/`dnxcore50`), then just make your assemblies and tests target desktop .NET.

**Comment by cherrydev** [comment]
My projects are targeting `aspnet50` and they have access to all of the desktop .NET framework.  They do NOT if you are targeting `aspnetcore50`.
I'm not quite sure I understand, are you saying that if I'm just going to be using `aspnet50/dnx451` anyway, I shouldn't bother with it and stick to writing .csproj projects and target .NET 4.5.1 instead?  That would lock me out of being able to use any of the new versions of the libraries, such as EF7, MVC6, etc… Have I misunderstood what you mean?
If you look into how MS has architected their ASPNET5 libraries, you'll see that some of their libraries have different versions depending on if they're targeting `aspnet50` or `aspnetcore50` (`Kestrel` is an example).  So, there absolutely is a straightforward way to support this and there is precedent and examples of it.
Some developers are going to go the route of using `aspnetcore50` so that they can develop portable applications.  Others will use `aspnet50` on Windows because that's how you get access to the new frameworks and tooling.
* Currently, the `xunit.runner.aspnet` runner is the only real way to run tests for the new .kproj/.xproj projects.
* AppDomain isolation is unquestionably a desirable feature in a test runner.
* `aspnet50` is a first class citizen in the new MS ecosystem.
* It is possible to publish `xunit.runner.aspnet` so that it performs AppDomain isolation when running on `aspnet50` and to perform as-is when running on `aspnetcore50`.
* You can make separate 'defines' inside of the `frameworks.<runtimeName>` nodes in project.json (though it's a shame that you can't currently specify separate `code` lists).

I hope this is enough to convince you that:
* It's possible to provide this support
* It's desirable to provide this support
* It's probably straightforward to provide this support

**Comment by cherrydev** [comment]
[Here](https://github.com/aspnet/EntityFramework/blob/master/test/EntityFramework.Commands.FunctionalTests/TestUtilities/ExecutorWrapper.cs) is an example of MS using AppDomains in their test harness for EF7.
[This](https://github.com/aspnet/SignalR-Server/blob/master/src/Microsoft.AspNet.SignalR.Server/Hubs/TypedClientBuilder.cs) is an example of using AppDomain to load assemblies.

**Comment by cherrydev** [comment]
And btw, my reasoning for not wanting to make due with manually creating AppDomains inside of my tests is because I would essentially have to replicate large portions of the XUnit framework in order to support things such as test fixtures that are created inside of the test's app domain.

**Comment by bradwilson** [comment]
I'm saying that our single code base for the ASP.NET runner will not be supporting App Domains. If you want App Domains, your tests must target the desktop CLR and use the desktop CLR runner.

You're welcome to fork the runner and add App Domains in for your own code, but we have no plans on supporting this scenario.

**Comment by cherrydev** [comment]
Okay.  My concern, though, is that the aspnet execution and runner are tying together two things that are independent:
* The development environment that the runner is being used in, in this case .kproj/.xproj style projects
* The runtime (of which many are supported in .kproj/.xproj projects)
As far as I know, there is no way to launch the desktop runner inside of a .kproj/.xproj project unless you were to set up the dev environment to produce .dll files each and every time, which doesn't fit well with the dev environment (and then I believe you'd need to manually set up the paths to the dependent assemblies).  If the desktop runner can only operate on a .csproj style project, it's actually impossible to use this project to test code inside of a .kproj/.xproj project, since there's no method of referencing a .kproj/.xproj project from a .csproj project.
If a straightforward PR came in that supported AppDomain isolation on runtimes that have it, would you consider it, or is this 100% a dead end?

**Comment by bradwilson** [comment]
There are, fundamentally, two reasons why we don't want conditional App Domain code in the ASP.NET 5 runner.

First: App Domains are the past. Relying on App Domains when writing ASP.NET 5 code seems like you're just begging to be stuck in the past. Moreover, since multiple assemblies can't be run in parallel with the ASP.NET runner at the moment, there really is no isolation value. The only thing you're really giving up is shadow copying (that has some benefit for some users).

Second: We hate `#if`. We accept that it's a necessary evil when writing something that targets 10 platforms, but we actively look for ways to reduce this cognitive burden. Without question, you would be adding to the load, not reducing it, by asking for this feature.

As a bonus, I still have no idea why you absolutely 100% **need** to have App Domains. What is the requirement that makes a lack of App Domains a complete non-starter for you? By this, I mean it has be a complete show-stopper; I'm not interested in simple preference.

**Comment by cherrydev** [comment]
Thanks for your explanation.
I'm not using AppDomains in my code at all.  To clarify, my ONLY concern is about the test runner being able to provide isolation between the assemblies that it's testing.  The only way that I know of to provide this isolation without running each assembly's test runner in a separate process is with AppDomains.  I know there's some sort of dynamic assembly loader that's specific to `aspnetcore50` but I don't know if it provides the same isolation functionality, and it requires `#if` statements to use anyway since that API is not actually available if you're using the `aspnet50` runtime.

Without this isolation, it is possible that errant 3rd party libraries will behave unpredictably/undesirably.  One specific example I have is EF6, which has a bunch of global state that is initialized the first time you create a DbContext and cannot be cleared.  This state includes the DbConfiguration instance and has always been one of the pain points in writing tests that involve Entity Framework. (The fact that I'm still working with EF6 in an aspnet5 project is another story…)

If you are adamant about having a single code path (which would certainly be easiest) for test execution that runs inside of a .xproj/.kproj dev environment, regardless of the runtime, I can make due without it.  I just think it would be preferable to developers using xunit if their tests assemblies could be run in isolation.  I understand that's contrary to your desire, as the maintainer or the project, to keep things as simple as possible.

Another possible idea that would not add a single `#if` statement: Make a `xunit.runner.utility.AspNetFullCLR` project that targets only aspnet5 (not aspnetcore5), swap out `LongLivedMarshalByRefObject_Runner_NoAppDomain.cs` and `RemoteAppDomainManager_NoAppDomain.cs` for their AppDomain counterparts and alter `xunit.runner.aspnet` `project.json` to depend on `xunit.runner.utility.AspNet` for `aspnetcore5` and `xunit.runner.utility.AspNetFullCLR` for `aspnet5`.  No new code, no new `#if` statements, just one more `project.json` file.  I may try this and see if's as simple as it sounds.

**Comment by bradwilson** [comment]
With the ASP.NET 5 runner, it's impossible to run tests from multiple assemblies at the same time in the same process (unlike the other runners), due to limitations in the way k/dnx invocation works. The only isolation you would gain would be from the runner itself, which while valuable, is certainly not necessary.

The only form of assembly parallelization for ASP.NET 5 test runs is to actually invoke the runner multiple times in separate processes, and those would absolutely be isolated from one another (arguably much better, since they'd be isolated by processes rather than app domains).

So I see really no need for this feature.

**Comment by cherrydev** [comment]
Well, you can run tests from multiple assemblies if you make a parent project that depends on each of them.  You can then specify the assembly names (of the dependent test assemblies) for the tests that you want to run on the command line.  That works fine right now, and is how I performed my tests for isolation. The limitation of having to use a parent project is, of course, also an issue that I brought up, but I don't actually believe that it's an inherent limitation of the dnx invocation, just that loading assemblies dynamically from a project folder uses new APIs.
Once that issue is taken care of, I hope you might reconsider this one.  I'd like to eventually see aspnet5 support on xUnit to be at least on par with other platforms.

**Comment by bradwilson** [comment]
Closing, as there are no plans to support this.
